### ISSUE TITLE
Unable to install

### ISSUE THREAD

**Issue by cberthiaume** [data-forge/data-forge-js#9]
I have a VueJS project built with [vue-cli](https://github.com/vuejs/vue-cli) that I'm struggling to use data-forge-js in. I did the following:

`npm install --save data-forge`

and added this import to my component:

`import dataForge from 'data-forge';`

In my component I attempt to create a simple DataFrame:

`const dataFrame = new dataForge.DataFrame({
        columnNames: ['Col1'],
        values: [
          [1],
          [5],
          [10],
        ],
      });`

Upon executing `npm run dev` I see the following errors and warning (there are many, I've posted a subset):

```
WARNING in ./node_modules/require_optional/index.js
82:18-42 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/require_optional/index.js
 @ ./node_modules/mongodb-core/lib/auth/gssapi.js
 @ ./node_modules/mongodb-core/index.js
 @ ./node_modules/promised-mongo/index.js
 @ ./node_modules/data-forge/src/dataframe.js
 @ ./node_modules/data-forge/src/series.js
 @ ./node_modules/data-forge/src/concat-series.js
 @ ./node_modules/data-forge/index.js
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/TrainerHome.vue
 @ ./src/components/TrainerHome.vue
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/MemberHome.vue
 @ ./src/components/MemberHome.vue
 @ ./src/router/index.js
 @ ./src/main.js
 @ multi ./build/dev-client ./src/main.js

WARNING in ./node_modules/require_optional/index.js
90:20-44 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/require_optional/index.js
 @ ./node_modules/mongodb-core/lib/auth/gssapi.js
 @ ./node_modules/mongodb-core/index.js
 @ ./node_modules/promised-mongo/index.js
 @ ./node_modules/data-forge/src/dataframe.js
 @ ./node_modules/data-forge/src/series.js
 @ ./node_modules/data-forge/src/concat-series.js
 @ ./node_modules/data-forge/index.js
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/TrainerHome.vue
 @ ./src/components/TrainerHome.vue
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/MemberHome.vue
 @ ./src/components/MemberHome.vue
 @ ./src/router/index.js
 @ ./src/main.js
 @ multi ./build/dev-client ./src/main.js

...

ERROR in ./node_modules/harmony-proxy/index.js
Module not found: Error: Can't resolve './lib-cov' in 'C:\Users\chris\data\work\move2thrive\move2thrive_web\node_modules\harmony-proxy'
 @ ./node_modules/harmony-proxy/index.js 2:4-24
 @ ./node_modules/promised-mongo/index.js
 @ ./node_modules/data-forge/src/dataframe.js
 @ ./node_modules/data-forge/src/series.js
 @ ./node_modules/data-forge/src/concat-series.js
 @ ./node_modules/data-forge/index.js
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/TrainerHome.vue
 @ ./src/components/TrainerHome.vue
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/MemberHome.vue
 @ ./src/components/MemberHome.vue
 @ ./src/router/index.js
 @ ./src/main.js
 @ multi ./build/dev-client ./src/main.js

ERROR in ./node_modules/babyparse/babyparse.js
Module not found: Error: Can't resolve 'fs' in 'C:\Users\chris\data\work\move2thrive\move2thrive_web\node_modules\babyparse'
 @ ./node_modules/babyparse/babyparse.js 41:16-29
 @ ./node_modules/data-forge/src/dataframe.js
 @ ./node_modules/data-forge/src/series.js
 @ ./node_modules/data-forge/src/concat-series.js
 @ ./node_modules/data-forge/index.js
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/TrainerHome.vue
 @ ./src/components/TrainerHome.vue
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/MemberHome.vue
 @ ./src/components/MemberHome.vue
 @ ./src/router/index.js
 @ ./src/main.js
 @ multi ./build/dev-client ./src/main.js
....
```
I've been able to install and import many other npm packages without an issue. I've done a lot of searching for these errors and warnings but, thus far, I've been unable to get it to work. I'd really like use data-forge-js in my project. Any assistance provided would be much appreciated.

**Comment by ashleydavis** [comment]
Hi, thanks for logging an issue.

I don't recognise the problem.  Are you able to provide an example that I can debug? If you have your project on Github already please give me a link to it.

**Comment by cberthiaume** [comment]
Hey, thanks for the fast response! Let me see if I can construct an example. My project is in gitlab and is private. I could potentially get you access if I can't create an example. Thanks again!

**Comment by ashleydavis** [comment]
Even just a small example and I can have a look at it today.

**Comment by cberthiaume** [comment]
I'll see what I can put together this evening (my time) after putting my kid to bed :)

**Comment by cberthiaume** [comment]
If you have a gitlab account I'd be open to granting you access to the repo there.

**Comment by ashleydavis** [comment]
Sorry haven't signed up for it.

**Comment by cberthiaume** [comment]
No problem, I'll work on getting you an example or access to the git
repository. Thank you again!

**Comment by cberthiaume** [comment]
I created a quick, if not simple, example that demonstrates the problem.
https://github.com/cberthiaume/data-forge-js-test

I started by using vue-cli version 2.9.3 to create a default project:

`vue init webpack data-forge-js-test`

Following the instructions in the README I can load the default page locally by running `npm run dev`. Next, I added the reference to data-forge and created a DataFrame in the `mounted` lifecycle hook. Now, when I run `npm run dev` I receive this error:

```
Failed to compile.

./node_modules/harmony-proxy/index.js
Module not found: Error: Can't resolve './lib-cov' in 'C:\Users\chris\data\work\move2thrive\data-forge-js-test\node_modules\harmony-proxy'
 @ ./node_modules/harmony-proxy/index.js 2:4-24
 @ ./node_modules/promised-mongo/index.js
 @ ./node_modules/data-forge/index.js
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/HelloWorld.vue
 @ ./src/components/HelloWorld.vue
 @ ./src/router/index.js
 @ ./src/main.js
 @ multi (webpack)-dev-server/client?http://localhost:8081 webpack/hot/dev-server ./src/main.js
```

You can see the change I made to the default in [this commit](https://github.com/cberthiaume/data-forge-js-test/commit/1a3a5bcde37886e6857a8181b316bf1bdede5e2a).

I hope this helps. Let me know if I can add anything further.

**Comment by ashleydavis** [comment]
I think I've fixed your problem.

Before I publish though can I please get you to test it with your project?

First uninstall your existing version:

`npm uninstall --save data-forge
`
Then install the latest version directly from the repo:

`npm install --save https://github.com/data-forge/data-forge-js
`

Please let me know if it works for you now, if so I'll publish it.

**Comment by cberthiaume** [comment]
That did it! The application builds now and I can see the console output of the DataFrame.

Thank you so much for looking into this so quickly!

**Comment by ashleydavis** [comment]
Ok no worries. It's getting late here so I'll probably publish tomorrow. I'll let you know when it's done.

You might be interested to know that data-forge-ts (the TypeScript port) is on it's way. Please jump over to the new repo and add a star ;)

https://github.com/data-forge/data-forge-ts

**Comment by cberthiaume** [comment]
Done!

**Comment by cberthiaume** [comment]
I attempted to use the latest version directly from the repo in my actual project, as opposed to the example, and I'm seeing these errors:

```
ERROR in ./node_modules/babyparse/babyparse.js
Module not found: Error: Can't resolve 'fs' in 'C:\Users\chris\data\work\move2thrive\move2thrive_web\node_modules\babyparse'
 @ ./node_modules/babyparse/babyparse.js 41:16-29
 @ ./node_modules/data-forge/index.js
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/TrainerHome.vue
 @ ./src/components/TrainerHome.vue
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/MemberHome.vue
 @ ./src/components/MemberHome.vue
 @ ./src/router/index.js
 @ ./src/main.js
 @ multi ./build/dev-client ./src/main.js

ERROR in ./node_modules/data-forge/index.js
Module not found: Error: Can't resolve 'fs' in 'C:\Users\chris\data\work\move2thrive\move2thrive_web\node_modules\data-forge'
 @ ./node_modules/data-forge/index.js 213:14-27 239:14-27 278:13-26 294:13-26
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/TrainerHome.vue
 @ ./src/components/TrainerHome.vue
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/MemberHome.vue
 @ ./src/components/MemberHome.vue
 @ ./src/router/index.js
 @ ./src/main.js
 @ multi ./build/dev-client ./src/main.js

ERROR in ./node_modules/data-forge/src/dataframe.js
Module not found: Error: Can't resolve 'fs' in 'C:\Users\chris\data\work\move2thrive\move2thrive_web\node_modules\data-forge\src'
 @ ./node_modules/data-forge/src/dataframe.js 1038:13-26 1059:12-25 1089:13-26 1110:12-25
 @ ./node_modules/data-forge/index.js
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/TrainerHome.vue
 @ ./src/components/TrainerHome.vue
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./src/components/MemberHome.vue
 @ ./src/components/MemberHome.vue
 @ ./src/router/index.js
 @ ./src/main.js
 @ multi ./build/dev-client ./src/main.js
```
If you're interested I could send you a zip of the repo to email address. I'd prefer not to post it publicly though.

**Comment by ashleydavis** [comment]
That's strange. 

Can you think  of any difference been your example and your real app that would manifest the problem?

If you can make the problem happen with your example I'm sure I'll be able to fix it quickly.

**Comment by ashleydavis** [comment]
Feel free to email the repo to ashley@codecapers.com.au.

I won't share it and I'll delete it once wee have fixed the problem.

**Comment by cberthiaume** [comment]
I think I see the difference. In my example, which apparently was initialized with a different version of vue-cli than my project, webpack.base.conf.js includes this block:

```
node: {
    // prevent webpack from injecting useless setImmediate polyfill because Vue
    // source contains it (although only uses it if it's native).
    setImmediate: false,
    // prevent webpack from injecting mocks to Node native modules
    // that does not make sense for the client
    dgram: 'empty',
    fs: 'empty',
    net: 'empty',
    tls: 'empty',
    child_process: 'empty'
  },
```
Adding that to the same file in my actual project remedies the problem and the application seems to work correctly. It also lines up with what I see when searching for that error on the web.

Thanks again for all the help!

**Comment by ashleydavis** [comment]
No problem. Glad to see it sorted. I hope to publish the updated version today or tomorrow.

If you are interested in working with data in JavaScript, I'm actually I'm writing a book on the subject!

It's available now on Manning early access (to be printed in the coming months):

https://www.manning.com/books/data-wrangling-with-javascript

If you want to support my efforts please let your colleagues and networks know about the book!

**Comment by cberthiaume** [comment]
Thanks, I'll likely buy a copy myself and will spread the word!

**Comment by ashleydavis** [comment]
I've published version 0.6.0 of DataForge. Please uninstall the repo and reinstall from npm. Let me know if it still works as expected!

**Comment by cberthiaume** [comment]
0.6.0 working correctly. Thanks again!
